runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,165 @@
+local M = {}
+
+local semantic_tokens = setmetatable({}, {
```

Comment:
An alternative would be storing the semantic tokens on the client object itself, I'm a bit concerned generally that we are spreading/storing client specified state across all of these different files. When the client self-terminates it should be able to cleanup it's own state (including stale semantic tokens) in on_exit.

Reply:
So we would probably add a per-buffer object, where clients can store their per-buffer data. So that we can clean-up on client termination or on buffer close.

Reply:
Yes, I was thinking like `client.state.buffer.{semantic_tokens,codelens}`